src/db.py

```diff
@@ -0,0 +1,250 @@
+
+    def push_job_to_queue(self, job: JobSchema) -> None:
+        list_name = get_list_name(job.priority)
+        self.redis.lpush(list_name, job.id)
+
+    def get_random_job_by_priority(self) -> Optional[Tuple[JobQueue, JobId]]:
+        """ Tries to get a random job along with its queue """
+        for queue in JobQueue.available():
+            yara_jobs = self.redis.lrange(queue.name, 0, -1)
+            if yara_jobs:
+                return queue, JobId(random.choice(yara_jobs))
+        return None
+
+    def get_job(self, job: JobId) -> JobSchema:
+        data = self.redis.hgetall(job.key)
+        return JobSchema(
+            id=job.hash,
+            status=data.get("status", "ERROR"),
+            rule_name=data.get("rule_name", "ERROR"),
+            rule_author=data.get("rule_author", None),
+            raw_yara=data.get("raw_yara", "ERROR"),
+            submitted=data.get("submitted", 0),
+            priority=data.get("priority", "medium"),
+            files_processed=data.get("files_processed", 0),
+            total_files=data.get("total_files", 0),
+            iterator=data.get("iterator", None),
+            taint=data.get("taint", None),
+        )
+
+    def add_match(self, job: JobId, match: MatchInfo) -> None:
+        self.redis.rpush(job.meta_key, match.to_json())
+
+    def job_contains(self, job: JobId, ordinal: str, file_path: str) -> bool:
+        file_list = self.redis.lrange(job.meta_key, ordinal, ordinal)
+        return file_list and file_path == json.loads(file_list[0])["file"]
+
+    def create_search_task(
+        self,
+        rule_name: str,
+        rule_author: str,
+        raw_yara: str,
+        priority: Optional[str],
+        taint: Optional[str],
+    ) -> JobId:
+        job = JobId(
+            "".join(
+                random.SystemRandom().choice(
+                    string.ascii_uppercase + string.digits
+                )
+                for _ in range(12)
+            )
+        )
+        job_obj = {
+            "status": "new",
+            "rule_name": rule_name,
+            "rule_author": rule_author,
+            "raw_yara": raw_yara,
+            "submitted": int(time()),
+            "priority": priority,
+        }
+        if taint is not None:
+            job_obj["taint"] = taint
+
+        self.redis.hmset(job.key, job_obj)
+        self.redis.rpush("queue-search", job.hash)
+        return job
+
+    def get_job_matches(
+        self, job: JobId, offset: int, limit: int
+    ) -> MatchesSchema:
+        p = self.redis.pipeline(transaction=False)
+        p.hgetall(job.key)
+        p.lrange("meta:" + job.hash, offset, offset + limit - 1)
+        job, meta = p.execute()
+        return MatchesSchema(job=job, matches=[json.loads(m) for m in meta])
+
+    def run_command(self, command: str) -> None:
+        self.redis.rpush("queue-commands", command)
+
+    def get_task(self) -> Tuple[Optional[str], Optional[str]]:
```

Comment:
Wouldn't be better to return `Optional[Tuple[str, str]]`?

Reply:
It would, I just tried to avoid doing oto many changes (and also was lazy). Fixed now.